driver-core/src/main/com/mongodb/operation/OperationHelper.java

```diff
@@ -444,6 +526,76 @@ static void withConnection(final AsyncReadBinding binding, final AsyncCallableWi
         binding.getReadConnectionSource(errorHandlingCallback(new AsyncCallableWithConnectionAndSourceCallback(callable), LOGGER));
     }
 
+    static <T> T retryableWithConnection(final WriteBinding binding, final boolean retryable,
+                                         final CallableWithConnectionAndTransactionNumber<T> callable) {
+        final Long txnNumber = retryable ? binding.getSessionContext().advanceTransactionNumber() : null;
+        return withConnection(binding, new CallableWithConnection<T>() {
```

Comment:
With nested calls to withConnection, the first connection is not released until after the second connection is checked out and released.  This could result in artificial connection pool exhaustion, since even if a network error forces the connection to be closed, the pool won't make room for a new one until the closed connection is released back to the pool.  Do something like this instead:

        try {
            return withConnection(binding, new CallableWithConnection<T>() {
                @Override
                public T call(Connection connection) {
                    return callable.call(connection, txnNumber);
                }
            });
        } catch (MongoException e) {
            if (!retryable) {
                throw e;
            } else {
                final MongoException originalException = e;
                 // NO LONGER NESTED
                return withConnection(binding, new CallableWithConnection<T>() {
                    @Override
                    public T call(final Connection connection) {
                        if (!serverIsAtLeastVersionThreeDotSix(connection.getDescription())) {
                            throw originalException;
                        }
                        return callable.call(connection, txnNumber);
                    }
                });
            }
        }